common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -550,6 +558,197 @@ public static UTF8String toTitleCase(final UTF8String target, final int collatio
+    boolean newWord = true;
+    // We are maintaining if the current character is preceded by a cased letter.
+    // This is used when lowercasing capital Greek letter sigma ('Σ'), to figure out if it should be
+    // lowercased into σ or ς.
+    boolean precededByCasedLetter = false;
+
+    // 'offset' is a byte offset in target's byte array pointing to the beginning of the character
+    // that we need to process next.
+    int offset = 0;
+    int len = target.numBytes();
+
+    while(offset < len) {
+      // We will actually call 'codePointFrom()' 2 times for each character in the worst case (once
+      // here, and once in 'followedByCasedLetter'). Example of a string where we call it 2 times
+      // for almost every character is 'ΣΣΣΣΣ' (a string consisting only of Greek capital sigma)
+      // and 'Σ`````' (a string consisting of a Greek capital sigma, followed by case-ignorable
+      // characters).
+      int codepoint = target.codePointFrom(offset);
+      // Appending the correctly cased character onto 'sb'.
+      appendTitleCasedCodepoint(sb, codepoint, newWord, precededByCasedLetter, target, offset);
+      // Updating 'newWord', 'precededByCasedLetter' and 'offset' to be ready for the next character
+      // that we will process.
+      newWord = (codepoint == ASCII_SPACE_CODEPOINT);
+      if(!UCharacter.hasBinaryProperty(codepoint, UProperty.CASE_IGNORABLE)) {
+        precededByCasedLetter = UCharacter.hasBinaryProperty(codepoint, UProperty.CASED);
+      }
+      offset += UTF8String.numBytesForFirstByte(target.getByte(offset));
+    }
+    return sb.build();
+  }
+
+  private static void appendTitleCasedCodepoint(
+      UTF8StringBuilder sb,
+      int codepoint,
+      boolean isAfterAsciiSpace,
+      boolean precededByCasedLetter,
+      UTF8String target,
+      int offset) {
+    if(isAfterAsciiSpace) {
+      // Titlecasing a character if it is in the beginning of a new word.
+      appendCodepointToTitleCase(sb, codepoint);
+      return;
+    }
+    if(codepoint == CAPITAL_SIGMA) {
+      // Handling capital Greek letter sigma ('Σ').
+      appendLowerCasedGreekCapitalSigma(sb, precededByCasedLetter, target, offset);
+      return;
+    }
+    // If it's not the beginning of a word, or a capital Greek letter sigma ('Σ'), we lowercase the
+    // character. We specially handle 'CAPITAL_I_WITH_DOT_ABOVE'.
+    if (codepoint == CAPITAL_I_WITH_DOT_ABOVE) {
+      sb.appendCodePoint(LOWERCASE_I);
+      sb.appendCodePoint(DOT_ABOVE);
+      return;
+    }
+    sb.appendCodePoint(UCharacter.toLowerCase(codepoint));
+  }
+
+  private static void appendLowerCasedGreekCapitalSigma(
+      UTF8StringBuilder sb,
+      boolean precededByCasedLetter,
+      UTF8String target,
+      int offset) {
+    int codepoint;
+    if (!followedByCasedLetter(target,offset) && precededByCasedLetter) {
+      codepoint = SMALL_FINAL_SIGMA;
+    }
+    else {
+      codepoint = SMALL_NON_FINAL_SIGMA;
+    }
```

Comment:
```suggestion
    int codepoint;
    if (!followedByCasedLetter(target, offset) && precededByCasedLetter) {
      codepoint = SMALL_FINAL_SIGMA;
    } else {
      codepoint = SMALL_NON_FINAL_SIGMA;
    }
```
Coding style. Please notice the space I added after the comma when calling `followedByCasedLetter`.

Reply:
Alternatively:
```java
  int codepoint = (!followedByCasedLetter(target, offset) && precededByCasedLetter) ? GREEK_FINAL_SIGMA : GREEK_SMALL_SIGMA;

Reply:
The initial implementation was with the '?' operator, but that row was over 100 characters long and i wasn't sure on how to format it so that it follows the coding style, so i sticked to the 'if' statement. I might have missed it in the code style doc. Could you show a way to format it properly when it doesn't fit a single row?

Reply:
As we discussed on Slack: we don't really have some tool for Java code formatting. My suggestion is to avoid lines that are more than 100 characters long use some online formatting tool at your discretion to produce some reasonably looking readable code.